### PR TITLE
feat(web/applications): key dates step

### PR DESCRIPTION
**Pull request by lutamKC**
## Describe your changes

Add UI for the key dates step + validators

## Type of change 🧩

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update
- [ ] Other (please explain in the description section above)

## Checklist before requesting a review

- [ ] I have performed a self-review of my own code
- [ ] I have double checked this work does not include any hardcoded secrets or passwords
- [ ] I have made corresponding changes to the documentation
- [ ] I have provided details on how I have tested my code
- [ ] I have referenced the ticket number above
- [ ] I have provided a description of the ticket
- [ ] I have included unit tests to cover any testable code changes

### CODE REVIEW

File: apps/web/src/server/applications/create/key-dates/__tests__/applications-create-key-dates.test.js [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+			it('should show validation errors if dates are not valid', async () => {
+				const response = await request.post(baseUrl).send({
+					firstNotifiedDay: 'not_a_number',
+					submissionDay: 32
+				});
+				const element = parseHtml(response.text);
+
+				expect(element.innerHTML).toMatchSnapshot();
+				expect(element.innerHTML).toContain('Enter a valid day for the first notified date');
+				expect(element.innerHTML).toContain('Enter a valid day for the submission date');
```
Comment (mayya-bond-kc) [inline apps/web/src/server/applications/create/key-dates/__tests__/applications-create-key-dates.test.js:39]: Is the error messaging "enter a valid day", when we are expecting a string that might just contain the quarter or something?

File: apps/web/src/server/applications/create/key-dates/__tests__/applications-create-key-dates.test.js [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+				const response = await request.post(baseUrl).send({
+					submissionDay: '14',
+					submissionMonth: '7',
+					submissionYear: '1789'
+				});
+				const element = parseHtml(response.text);
+
+				expect(element.innerHTML).toMatchSnapshot();
+				expect(element.innerHTML).toContain('The submission date should be in the future');
+			});
```
Comment (mayya-bond-kc) [inline apps/web/src/server/applications/create/key-dates/__tests__/applications-create-key-dates.test.js:52]: Again, thought the submission date is now a string without any specific format?

File: apps/web/src/server/applications/create/key-dates/applications-create-key-dates.controller.js [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+		firstNotifiedDay,
+		firstNotifiedMonth,
+		firstNotifiedYear,
+		submissionDay,
+		submissionMonth,
+		submissionYear
+	} = body;
+
+	const firstNotifiedSeconds =
+		(Date.parse(`${firstNotifiedYear}-${firstNotifiedMonth}-${firstNotifiedDay}`) || 0) / 1000;
```
Comment (mayya-bond-kc) [inline apps/web/src/server/applications/create/key-dates/applications-create-key-dates.controller.js:46]: Why not sure here `new Date(firstNotifiedYear, firstNotifiedMonth, firstNotifiedDay)`?

File: apps/web/src/server/applications/create/key-dates/applications-create-key-dates.controller.js [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+	const submissionDateSeconds =
+		(Date.parse(`${submissionYear}-${submissionMonth}-${submissionDay}`) || 0) / 1000;
```
Comment (mayya-bond-kc) [inline apps/web/src/server/applications/create/key-dates/applications-create-key-dates.controller.js:48]: Same question around submission date

File: apps/web/src/server/lib/dates.js [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+ */
+export const isDateInstance = (date) => {
+	return (
+		date instanceof Date &&
+		!Number.isNaN(date) &&
+		!Number.isNaN(date.getDate()) &&
+		!Number.isNaN(date.getMonth()) &&
+		!Number.isNaN(date.getFullYear())
+	);
+};
```
Comment (mayya-bond-kc) [inline apps/web/src/server/lib/dates.js:32]: I did not realise how much effort it is to validate a date

**Review by mayya-bond-kc — COMMENTED** [review]
Looks good. Main question is around the submission date being a Date VS String

File: apps/web/src/server/lib/dates.js (hunk 5 above)
Comment (lutamKC) [reply apps/web/src/server/lib/dates.js:32]: last two are not really necessary

File: apps/web/src/server/applications/create/key-dates/__tests__/applications-create-key-dates.test.js (hunk 1 above)
Comment (lutamKC) [reply apps/web/src/server/applications/create/key-dates/__tests__/applications-create-key-dates.test.js:39]: ok. apparently there a two designs versions for this page, the correct one that you're talking about and the other.
guess what I was looking at?

File: apps/web/src/server/lib/dates.js (hunk 5 above)
Comment (lutamKC) [reply apps/web/src/server/lib/dates.js:32]: but we can think of using Moment.js or something like that to make dates manipulations easier

File: apps/web/src/server/applications/create/key-dates/__tests__/applications-create-key-dates.test.js (hunk 2 above)
Comment (lutamKC) [reply apps/web/src/server/applications/create/key-dates/__tests__/applications-create-key-dates.test.js]: submission date (internal) is still a date

File: apps/web/src/server/applications/create/key-dates/applications-create-key-dates.controller.js (hunk 4 above)
Comment (lutamKC) [reply apps/web/src/server/applications/create/key-dates/applications-create-key-dates.controller.js]: same as ⬆️.
i changed submission date published (former publishingdate) to be a generic string but submssion date internal keeps being a date with days,months and years

**Review by mayya-bond-kc — APPROVED** [review]

**Pull request merged**